apollo-federation/src/merger/merge_input.rs

```diff
@@ -0,0 +1,605 @@
+use apollo_compiler::schema::InputObjectType;
+
+use crate::error::CompositionError;
+use crate::error::FederationError;
+use crate::error::SubgraphLocation;
+use crate::merger::hints::HintCode;
+use crate::merger::merge::Merger;
+use crate::merger::merge::Sources;
+use crate::schema::position::InputObjectTypeDefinitionPosition;
+use crate::supergraph::CompositionHint;
+
+impl Merger {
+    #[allow(dead_code)]
+    pub(crate) fn merge_input(
+        &mut self,
+        sources: &Sources<Node<InputObjectType>>,
+        dest: &InputObjectTypeDefinitionPosition,
+    ) -> Result<(), FederationError> {
+        let added = self.add_fields_shallow(sources, dest);
+
+        for (dest_field, subgraph_fields) in added {
+            if let Some(field_def) = subgraph_fields.values().find_map(|f| f.as_ref()) {
+                if dest_field.try_get(self.merged.schema()).is_none() {
+                    dest_field.insert(&mut self.merged, field_def.clone())?;
+                }
+            }
+
+            let dest_field_def = dest_field.try_get(self.merged.schema());
+            let is_inaccessible = if let Some(dest_field_def) = dest_field_def {
+                if let Some(inaccessible_directive) =
+                    &self.inaccessible_directive_name_in_supergraph
+                {
+                    dest_field_def.directives.has(inaccessible_directive)
+                } else {
+                    false
+                }
+            } else {
+                false
+            };
+
+            // Note: if the field is marked @inaccessible, we can always accept it to be inconsistent between subgraphs since
+            // it won't be exposed in the API, and we don't hint about it because we're just doing what the user is explicitly asking.
+            if !is_inaccessible
+                && Self::some_sources(&subgraph_fields, |field, _idx| field.is_none())
+            {
+                // If the field is optional, we remove it for the supergraph and issue a hint.
+                let mut non_optional_subgraphs = Vec::new();
+                let mut missing_subgraphs = Vec::new();
+
+                for (idx, field) in subgraph_fields.iter() {
+                    let subgraph_name = &self.names[*idx];
+                    match field {
+                        Some(field_def) if field_def.is_required() => {
+                            non_optional_subgraphs.push(subgraph_name);
+                        }
+                        None => {
+                            missing_subgraphs.push(subgraph_name);
+                        }
+                        _ => {}
+                    }
+                }
+
+                if !non_optional_subgraphs.is_empty() {
+                    let non_optional_subgraphs_str = non_optional_subgraphs
+                        .iter()
+                        .map(|s| s.as_str())
+                        .collect::<Vec<_>>()
+                        .join(", ");
```

Comment:
You don't need to allocate a new vector. `Itertools` provides a methods for this called `join`. Alternatively, you can skip allocating a new string altogether and use `Itertools::format`. Both of these would work
```rust
// Option A
let non_optional_subgraphs_str = non_optional_subgraphs.into_iter().join(", ");
// Option B
let non_optional_subgraphs_str = non_optional_subgraphs.into_iter().format(", ");
```